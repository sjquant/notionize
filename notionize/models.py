from pydantic import BaseModel, ConfigDict, Field, model_serializer
from typing import Any

from notionize.enums import NotionBlockType, NotionColor


class NotionModel(BaseModel):
    """Base model class for all Notion-related data structures with strict validation."""

    model_config = ConfigDict(frozen=True, extra="forbid")


class NotionAnnotations(NotionModel):
    """Text formatting properties for Notion rich text, including bold, italic, and color options."""

    bold: bool = False
    italic: bool = False
    strikethrough: bool = False
    underline: bool = False
    code: bool = False
    color: NotionColor = NotionColor.DEFAULT


class NotionLink(NotionModel):
    """Represents a URL link in Notion content with validation for URL format."""

    url: str


class NotionTextContent(NotionModel):
    """Container for text content and its associated link (if any) in Notion blocks."""

    content: str
    link: NotionLink | None = None


class NotionRichText(NotionModel):
    """Represents a rich text segment in Notion with formatting annotations and content."""

    type: str = "text"
    text: NotionTextContent
    annotations: NotionAnnotations = Field(default_factory=NotionAnnotations)


class NotionExternal(NotionModel):
    """Represents an external image in Notion with a URL."""

    url: str


class NotionImage(NotionModel):
    """Represents an image in Notion with a URL and optional caption."""

    type: str = "image"
    external: NotionExternal


class NotionBlock(BaseModel):
    """Base class for Notion block types with serialization support for API compatibility."""

    object: str = "block"
    type: NotionBlockType
    content: dict[str, Any]

    @model_serializer
    def to_api_format(self) -> dict[str, Any]:
        return {
            "object": self.object,
            "type": self.type.value,
            self.type.value: self.content,
        }

    model_config = ConfigDict(frozen=True, extra="forbid")
