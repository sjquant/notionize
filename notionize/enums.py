from enum import Enum


class NotionColor(str, Enum):
    DEFAULT = "default"
    GRAY = "gray"
    BROWN = "brown"
    ORANGE = "orange"
    YELLOW = "yellow"
    GREEN = "green"
    BLUE = "blue"
    PURPLE = "purple"
    PINK = "pink"
    RED = "red"
    GRAY_BACKGROUND = "gray_background"
    BROWN_BACKGROUND = "brown_background"
    ORANGE_BACKGROUND = "orange_background"
    YELLOW_BACKGROUND = "yellow_background"
    GREEN_BACKGROUND = "green_background"
    BLUE_BACKGROUND = "blue_background"
    PURPLE_BACKGROUND = "purple_background"
    PINK_BACKGROUND = "pink_background"
    RED_BACKGROUND = "red_background"


class NotionBlockType(str, Enum):
    PARAGRAPH = "paragraph"
    HEADING_1 = "heading_1"
    HEADING_2 = "heading_2"
    HEADING_3 = "heading_3"
    BULLETED = "bulleted_list_item"
    NUMBERED = "numbered_list_item"
    TO_DO = "to_do"
    TOGGLE = "toggle"
    CHILD_PAGE = "child_page"
    CHILD_DATABASE = "child_database"
    EMBED = "embed"
    IMAGE = "image"
    VIDEO = "video"
    FILE = "file"
    PDF = "pdf"
    BOOKMARK = "bookmark"
    CALLOUT = "callout"
    QUOTE = "quote"
    CODE = "code"
    DIVIDER = "divider"
    TABLE_OF_CONTENTS = "table_of_contents"
    COLUMN = "column"
    COLUMN_LIST = "column_list"
    LINK_PREVIEW = "link_preview"
    SYNCED_BLOCK = "synced_block"
    TEMPLATE = "template"
    LINK_TO_PAGE = "link_to_page"
    TABLE = "table"
    TABLE_ROW = "table_row"
    EQUATION = "equation"


class NotionLanguage(str, Enum):
    PLAIN_TEXT = "plain text"
    PYTHON = "python"
    JAVASCRIPT = "javascript"
    TYPESCRIPT = "typescript"
    JAVA = "java"
    CPP = "cpp"
    CSS = "css"
    HTML = "html"
    GO = "go"
    RUST = "rust"
    SHELL = "shell"
    SQL = "sql"
    YAML = "yaml"
