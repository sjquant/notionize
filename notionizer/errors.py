class NotionizerError(Exception):
    pass


class InvalidMarkdownError(NotionizerError):
    pass


class UnknownTokenError(NotionizerError):
    pass


class ConversionError(NotionizerError):
    pass
