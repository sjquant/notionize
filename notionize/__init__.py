from notionize.main import notionize, Notionizer


__all__ = ["Notionizer", "notionize"]
