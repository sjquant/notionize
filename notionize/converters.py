from dataclasses import dataclass
from typing import Any, Literal, cast
from abc import ABC, abstractmethod

from notionize.enums import NotionBlockType, NotionLanguage
from notionize.models import (
    NotionAnnotations,
    NotionBlock,
    NotionExternal,
    NotionImage,
    NotionLink,
    NotionModel,
    NotionRichText,
    NotionTextContent,
)


class BlockConverter(ABC):
    """Abstract base class for converting markdown tokens into Notion blocks.
    All concrete converter classes must implement the convert method."""

    @abstractmethod
    def convert(self, token: dict[str, Any]) -> NotionBlock | list[NotionBlock] | None:
        pass


class ParagraphConverter(BlockConverter):
    """Converts markdown paragraph tokens into Notion paragraph blocks."""

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        blocks = convert_inline_tokens(token.get("children", []))
        return NotionBlock(
            type=NotionBlockType.PARAGRAPH, content={"rich_text": blocks}
        )


class HeadingConverter(BlockConverter):
    """Converts markdown heading tokens into Notion heading blocks (h1-h3)."""

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        level = self._get_level(token)
        heading_type = getattr(NotionBlockType, f"HEADING_{level}")
        rich_text = convert_inline_tokens(token.get("children", []))
        return NotionBlock(type=heading_type, content={"rich_text": rich_text})

    def _get_level(self, token: dict[str, Any]) -> int:
        level = token.get("level")
        res = None
        if level := token.get("level"):
            res = level
        elif attrs := token.get("attrs"):
            res = attrs.get("level")

        if res is None:
            return 1

        if not 1 <= res <= 3:
            return 3

        return res


class CodeBlockConverter(BlockConverter):
    """Converts markdown code block tokens into Notion code blocks with language detection."""

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        language = self._get_lang(token)

        return NotionBlock(
            type=NotionBlockType.CODE,
            content={
                "rich_text": [
                    NotionRichText(text=NotionTextContent(content=token.get("raw", "")))
                ],
                "language": language,
            },
        )

    def _get_lang(self, token) -> str:
        try:
            if lang := token.get("info"):
                return lang
            elif attrs := token.get("attrs"):
                return attrs.get("info", NotionLanguage.PLAIN_TEXT)
            else:
                return NotionLanguage.PLAIN_TEXT
        except ValueError:
            return NotionLanguage.PLAIN_TEXT


class ListConverter(BlockConverter):
    """Converts markdown list tokens into Notion list blocks."""

    def convert(self, token: dict[str, Any]) -> list[NotionBlock]:
        res: list[NotionBlock] = []

        for item in token.get("children", []):
            converter = ListItemConverter(
                list_type=(
                    NotionBlockType.BULLETED
                    if token.get("ordered", False)
                    else NotionBlockType.NUMBERED
                )
            )
            res.append(converter.convert(item))

        return res


class ListItemConverter(BlockConverter):
    """Converts markdown list item tokens into Notion bulleted or numbered list items.
    Supports nested lists and mixed content within list items."""

    def __init__(
        self,
        list_type: Literal[NotionBlockType.BULLETED, NotionBlockType.NUMBERED],
    ):
        self.list_type = list_type

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        blocks: list[NotionModel] = []
        children_blocks: list[NotionBlock] = []

        for child in token.get("children", []):
            if child["type"] == "block_text" or child["type"] == "paragraph":
                blocks.extend(convert_inline_tokens(child.get("children", [])))
            else:
                # For Nested List
                from notionize.main import Notionizer

                notionizer = Notionizer()
                nested_blocks = notionizer.convert_blocks([child])
                if nested_blocks:
                    children_blocks.extend(nested_blocks)

        content: dict[str, Any] = {"rich_text": blocks}

        if children_blocks:
            content["children"] = children_blocks

        return NotionBlock(
            type=self.list_type,
            content=cast(dict[str, Any], content),
        )


class QuoteConverter(BlockConverter):
    """Converts markdown blockquote tokens into Notion quote blocks."""

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        blocks = convert_inline_tokens(token.get("children", []))
        return NotionBlock(type=NotionBlockType.QUOTE, content={"rich_text": blocks})


class DividerConverter(BlockConverter):
    """Converts markdown horizontal rule tokens into Notion divider blocks."""

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        return NotionBlock(type=NotionBlockType.DIVIDER, content={})


@dataclass(frozen=True, slots=True)
class TableStructure:
    header_cells: list[dict[str, Any]]
    body_rows: list[dict[str, Any]]


class TableConverter(BlockConverter):
    """Converts markdown table tokens into Notion table blocks.
    Supports tables with headers and handles special cases like link cells."""

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        table_structure = self._extract_table_structure(token)
        table_rows = self._process_table_rows(table_structure)

        return NotionBlock(
            type=NotionBlockType.TABLE,
            content={
                "table_width": self._calculate_table_width(table_structure),
                "has_column_header": bool(table_structure.header_cells),
                "has_row_header": False,
                "children": [
                    {"type": "table_row", "table_row": {"cells": row}}
                    for row in table_rows
                ],
            },
        )

    def _extract_table_structure(self, token: dict[str, Any]) -> TableStructure:
        table_head = next(
            (
                child
                for child in token.get("children", [])
                if child["type"] == "table_head"
            ),
            None,
        )
        table_body = next(
            (
                child
                for child in token.get("children", [])
                if child["type"] == "table_body"
            ),
            None,
        )

        header_cells = table_head.get("children", []) if table_head else []
        body_rows = table_body.get("children", []) if table_body else []

        return TableStructure(header_cells, body_rows)

    def _calculate_table_width(self, table_structure: TableStructure) -> int:
        if table_structure.header_cells:
            return len(table_structure.header_cells)
        if table_structure.body_rows:
            return len(table_structure.body_rows[0])
        return 1

    def _process_table_rows(
        self, table_structure: TableStructure
    ) -> list[list[list[NotionModel]]]:
        table_rows = []

        if table_structure.header_cells:
            table_rows.append(self._convert_row(table_structure.header_cells))

        for row in table_structure.body_rows:
            table_rows.append(self._convert_row(row.get("children", [])))

        return table_rows

    def _convert_row(self, cells: list[dict[str, Any]]) -> list[list[NotionModel]]:
        result = []

        for cell in cells:
            cell_content = cell.get("children", [])
            if self._is_link_cell(cell_content):
                result.append(
                    cast(list[NotionModel], self._convert_link_cell(cell_content[0]))
                )
            else:
                result.append(convert_inline_tokens(cell_content))
        return result

    def _is_link_cell(self, cell_content: list[dict[str, Any]]) -> bool:
        return bool(cell_content and cell_content[0].get("type") == "link")

    def _convert_link_cell(self, link_token: dict[str, Any]) -> list[NotionRichText]:
        return [
            NotionRichText(
                text=NotionTextContent(
                    content=link_token.get("children", [{}])[0].get("raw", ""),
                    link=NotionLink(url=link_token.get("attrs", {}).get("url", "")),
                )
            )
        ]


class ImageConverter(BlockConverter):
    """Converts markdown image tokens into Notion image blocks."""

    def convert(self, token: dict[str, Any]) -> NotionBlock:
        image_url = token.get("attrs", {}).get("url", "")
        return NotionBlock(
            type=NotionBlockType.IMAGE,
            content={"type": "external", "external": {"url": image_url}},
        )


class NullConverter(BlockConverter):
    """A null object pattern implementation that returns None for any conversion.
    Used as a fallback for unsupported markdown token types."""

    def convert(self, token: dict[str, Any]) -> None:
        return None


def convert_inline_tokens(tokens: list[dict[str, Any]]) -> list[NotionModel]:
    result: list[NotionModel] = []
    for token in tokens:
        if token["type"] == "text":
            result.append(create_text(token["raw"]))
        elif token["type"] in ("strong", "emphasis"):
            result.extend(create_formatted_texts(token))
        elif token["type"] == "link":
            result.extend(create_links(token))
        elif token["type"] == "image":
            result.extend(create_images(token))
    return result


def create_text(
    content: str,
    annotations: dict[str, Any] | None = None,
    link: str | None = None,
) -> NotionRichText:
    return NotionRichText(
        text=NotionTextContent(
            content=content, link=NotionLink(url=link) if link else None
        ),
        annotations=NotionAnnotations(**(annotations or {})),
    )


def create_formatted_texts(token: dict[str, Any]) -> list[NotionRichText]:
    annotations = {"bold": True} if token["type"] == "strong" else {"italic": True}
    result = []
    for child in token.get("children", []):
        if child["type"] == "text":
            result.append(create_text(child["raw"], annotations))
    return result


def create_links(token: dict[str, Any]) -> list[NotionRichText]:
    return [
        create_text(
            token.get("children", [{}])[0].get("text", ""),
            link=token.get("link", ""),
        )
    ]


def create_images(token: dict[str, Any]) -> list[NotionImage]:
    return [
        NotionImage(
            external=NotionExternal(url=token.get("attrs", {}).get("url", "")),
        )
    ]
