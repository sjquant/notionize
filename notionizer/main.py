from typing import Any, Callable
from notionizer.enums import NotionBlockType
from notionizer.converters import (
    BlockConverter,
    CodeBlockConverter,
    DividerConverter,
    HeadingConverter,
    ImageConverter,
    ListConverter,
    NullConverter,
    ParagraphConverter,
    QuoteConverter,
    TableConverter,
)
from notionizer.errors import ConversionError, InvalidMarkdownError, UnknownTokenError
from notionizer.models import NotionBlock
import mistune

ConverterFactory = Callable[[dict[str, Any]], BlockConverter | None]


class Notionizer:
    """A class that converts Markdown content into Notion API-compatible blocks.

    This class handles the parsing of Markdown content and conversion into Notion blocks
    using various converters for different Markdown elements (headings, paragraphs, lists, etc.).

    Args:
        converter_factory: Optional callable that creates custom block converters.
            It should take a token dict and return a BlockConverter or None.
    """

    def __init__(
        self,
        *,
        converter_factory: ConverterFactory | None = None,
    ):
        self._converter_factory = converter_factory

    def _get_converter(self, token: dict[str, Any]) -> BlockConverter:
        if self._converter_factory:
            res: BlockConverter | None = self._converter_factory(token)

            if res is not None:
                return res

        token_type = token.get("type")

        if token_type is None or token_type == "blank_line":
            return NullConverter()
        elif token_type == "paragraph" or token_type == "block_text":
            return ParagraphConverter()
        elif token_type == "heading":
            return HeadingConverter()
        elif token_type == "block_code":
            return CodeBlockConverter()
        elif token_type == "block_quote":
            return QuoteConverter()
        elif token_type == "thematic_break":
            return DividerConverter()
        elif token_type == "list" and token.get("ordered", False):
            return ListConverter(list_type=NotionBlockType.NUMBERED)
        elif token_type == "list" and not token.get("ordered", False):
            return ListConverter(list_type=NotionBlockType.BULLETED)
        elif token_type == "list_item":
            return ListConverter(list_type=NotionBlockType.BULLETED)
        elif token_type == "table":
            return TableConverter()
        elif token_type == "image":
            return ImageConverter()
        else:
            raise UnknownTokenError(
                f"Unknown token type: {token_type}. "
                "If you need to handle this token type, provide a custom converter_factory"
            )

    def run(self, content: str) -> list[dict[str, Any]]:
        """Convert Markdown content into a list of Notion blocks.

        Args:
            content: Markdown string to convert.

        Returns:
            List of dictionaries representing Notion blocks, ready for the Notion API.

        Raises:
            InvalidMarkdownError: If the markdown content cannot be parsed properly.
        """
        tokens = mistune.create_markdown(renderer="ast", plugins=["table"])(content)
        if isinstance(tokens, str):
            raise InvalidMarkdownError(
                "Failed to parse markdown: expected list of tokens but got string"
            )

        blocks = self.convert_blocks(tokens)
        return [block.model_dump(mode="json") for block in blocks]

    def convert_blocks(self, tokens: list[dict[str, Any]]) -> list[NotionBlock]:
        """Convert a list of Markdown tokens into Notion blocks.

        Args:
            tokens: List of Markdown tokens from the parser.

        Returns:
            List of NotionBlock objects.

        Raises:
            ConversionError: If conversion of any block fails.
        """
        res: list[NotionBlock] = []
        for token in tokens:
            token_type = token.get("type")

            if token_type is None:
                continue

            converter = self._get_converter(token)
            try:
                block = converter.convert(token)
                if block is not None:
                    res.append(block)

            except Exception as e:
                raise ConversionError(
                    f"Failed to convert token of type '{token_type}'",
                ) from e

        return res


def notionize(
    content: str,
    *,
    converter_factory: ConverterFactory | None = None,
) -> list[dict[str, Any]]:
    """Convert Markdown content into Notion API-compatible blocks.

    This is the main function for converting Markdown to Notion blocks. It creates
    a Notionizer instance and processes the content.

    Args:
        content: Markdown string to convert.
        converter_factory: Optional callable that creates custom block converters.
            It should take a token dict and return a BlockConverter or None.

    Returns:
        List of dictionaries representing Notion blocks, ready for the Notion API.

    Raises:
        InvalidMarkdownError: If the markdown content cannot be parsed properly.
    """
    notionizer = Notionizer(converter_factory=converter_factory)
    return notionizer.run(content)
