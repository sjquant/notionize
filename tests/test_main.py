import pytest
from notionize.enums import NotionBlockType
from notionize.main import Notionizer, notionize
from notionize.errors import ConversionError, UnknownTokenError
from notionize.converters import BlockConverter
from notionize.models import NotionBlock

from syrupy.assertion import SnapshotAssertion


def test_paragraph_conversion(snapshot: SnapshotAssertion):
    """Test paragraph conversion with various inline elements."""
    markdown = """This is a simple paragraph.
    
This has **bold** and _italic_ text and [a link](https://example.com)."""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_heading_conversion(snapshot: SnapshotAssertion):
    """Test heading conversion with all levels."""
    markdown = """# Heading 1
## Heading 2
### Heading 3
#### Heading 4 (should be converted to h3)"""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_code_block_conversion(snapshot: SnapshotAssertion):
    """Test code block conversion with different languages."""
    markdown = """```python
def hello():
    print("Hello")
```

```javascript
console.log('Hello');
```

```
Plain text code block
```"""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_list_conversion(snapshot: SnapshotAssertion):
    """Test list conversion with nested and mixed lists."""
    markdown = """- Unordered item 1
  - Nested unordered
    1. Nested ordered
    2. Another ordered
- Unordered item 2
  Some paragraph in list
  
1. Ordered item 1
2. Ordered item 2
   - Nested unordered
   - Another unordered"""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_quote_conversion(snapshot: SnapshotAssertion):
    """Test quote conversion with multiple lines and formatting."""
    markdown = """> This is a blockquote
> With **bold** and _italic_ text
> And multiple lines"""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_divider_conversion(snapshot: SnapshotAssertion):
    """Test divider converter."""
    markdown = """Some text

---

More text"""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_table_conversion(snapshot: SnapshotAssertion):
    """Test table conversion with headers, links, and formatting."""
    markdown = """| Header 1 | Header 2 | Header 3 |
|----------|----------|----------|
| Normal cell | **Bold** cell | [Link](https://example.com) |
| _Italic_ | Multiple<br>lines | Simple text |"""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_image_conversion(snapshot: SnapshotAssertion):
    """Test image conversion."""
    markdown = """![Image alt text](https://example.com/image.jpg)"""

    notionizer = Notionizer()
    res = notionizer.run(markdown)
    assert res == snapshot


def test_custom_converter_factory():
    """
    Test the ability to override default conversion behavior using a custom converter factory.
    """

    class CustomParagraphConverter(BlockConverter):
        def convert(self, token: dict) -> NotionBlock:
            return NotionBlock(
                type=NotionBlockType.PARAGRAPH,
                content={
                    "rich_text": [
                        {
                            "type": "text",
                            "text": {"content": "CUSTOM: " + token.get("text", "")},
                        }
                    ]
                },
            )

    def custom_factory(token: dict) -> BlockConverter | None:
        if token.get("type") == "paragraph":
            return CustomParagraphConverter()
        return None

    markdown = "This is a test paragraph."
    notionizer = Notionizer(converter_factory=custom_factory)
    res = notionizer.run(markdown)
    assert res == [
        {
            "object": "block",
            "type": "paragraph",
            "paragraph": {
                "rich_text": [{"type": "text", "text": {"content": "CUSTOM: "}}]
            },
        }
    ]


def test_unknown_token_type_raises_error():
    """
    Test that providing an unknown token type to convert_blocks directly raises a ValueError.
    """
    notionizer = Notionizer()
    unknown_token = {"type": "unknown_type", "text": "test"}

    try:
        notionizer.convert_blocks([unknown_token])
        assert False, "Expected UnknownTokenError"
    except UnknownTokenError as e:
        assert "Unknown token type: unknown_type" in str(e)


def test_conversion_error_raised():
    """
    Test that if the conversion fails, a ConversionError is raised.
    """

    class FailingConverter(BlockConverter):
        def convert(self, token: dict) -> NotionBlock:
            raise ValueError("Conversion failed")

    def failing_factory(token: dict) -> BlockConverter | None:
        return FailingConverter()

    notionizer = Notionizer(converter_factory=failing_factory)
    with pytest.raises(ConversionError) as exc_info:
        notionizer.run("Test paragraph")
    assert "Failed to convert token" in str(exc_info.value)


def test_notionize(snapshot: SnapshotAssertion):
    """
    Test the notionize function with comprehensive Markdown syntax.
    """
    markdown = """# Heading 1
## Heading 2
### Heading 3

This is a paragraph with **bold**, _italic_, and `inline code` text.

> This is a blockquote
> with multiple lines

1. Ordered list item 1
2. Ordered list item 2
   - Nested unordered item
   - Another nested item

- Unordered list item
- Another unordered item
  1. Nested ordered item
  2. Another nested ordered item

```python
def hello_world():
    print("Hello World!")
```

---

[Link text](https://example.com)

![Image alt text](https://example.com/image.jpg)

| Header 1 | Header 2 |
|----------|----------|
| Cell 1   | Cell 2   |
| Cell 3   | Cell 4   |
"""

    res = notionize(markdown)
    assert res == snapshot
