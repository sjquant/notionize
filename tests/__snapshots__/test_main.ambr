# serializer version: 1
# name: test_notionize
  list([
    dict({
      'heading_1': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'Heading 1',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'heading_1',
    }),
    dict({
      'heading_2': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'Heading 2',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'heading_2',
    }),
    dict({
      'heading_3': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'Heading 3',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'heading_3',
    }),
    dict({
      'object': 'block',
      'paragraph': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'This is a paragraph with ',
            }),
            'type': 'text',
          }),
          dict({
            'annotations': dict({
              'bold': True,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'bold',
            }),
            'type': 'text',
          }),
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': ', ',
            }),
            'type': 'text',
          }),
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': True,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'italic',
            }),
            'type': 'text',
          }),
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': ', and ',
            }),
            'type': 'text',
          }),
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': ' text.',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'type': 'paragraph',
    }),
    dict({
      'object': 'block',
      'quote': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'This is a blockquote',
            }),
            'type': 'text',
          }),
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'with multiple lines',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'type': 'quote',
    }),
    dict({
      'numbered_list_item': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'Ordered list item 1',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'numbered_list_item',
    }),
    dict({
      'numbered_list_item': dict({
        'children': list([
          dict({
            'bulleted_list_item': dict({
              'rich_text': list([
                dict({
                  'annotations': dict({
                    'bold': False,
                    'code': False,
                    'color': 'default',
                    'italic': False,
                    'strikethrough': False,
                    'underline': False,
                  }),
                  'text': dict({
                    'content': 'Nested unordered item',
                  }),
                  'type': 'text',
                }),
              ]),
            }),
            'object': 'block',
            'type': 'bulleted_list_item',
          }),
          dict({
            'bulleted_list_item': dict({
              'rich_text': list([
                dict({
                  'annotations': dict({
                    'bold': False,
                    'code': False,
                    'color': 'default',
                    'italic': False,
                    'strikethrough': False,
                    'underline': False,
                  }),
                  'text': dict({
                    'content': 'Another nested item',
                  }),
                  'type': 'text',
                }),
              ]),
            }),
            'object': 'block',
            'type': 'bulleted_list_item',
          }),
        ]),
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'Ordered list item 2',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'numbered_list_item',
    }),
    dict({
      'bulleted_list_item': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'Unordered list item',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'bulleted_list_item',
    }),
    dict({
      'bulleted_list_item': dict({
        'children': list([
          dict({
            'numbered_list_item': dict({
              'rich_text': list([
                dict({
                  'annotations': dict({
                    'bold': False,
                    'code': False,
                    'color': 'default',
                    'italic': False,
                    'strikethrough': False,
                    'underline': False,
                  }),
                  'text': dict({
                    'content': 'Nested ordered item',
                  }),
                  'type': 'text',
                }),
              ]),
            }),
            'object': 'block',
            'type': 'numbered_list_item',
          }),
          dict({
            'numbered_list_item': dict({
              'rich_text': list([
                dict({
                  'annotations': dict({
                    'bold': False,
                    'code': False,
                    'color': 'default',
                    'italic': False,
                    'strikethrough': False,
                    'underline': False,
                  }),
                  'text': dict({
                    'content': 'Another nested ordered item',
                  }),
                  'type': 'text',
                }),
              ]),
            }),
            'object': 'block',
            'type': 'numbered_list_item',
          }),
        ]),
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': 'Another unordered item',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'bulleted_list_item',
    }),
    dict({
      'code': dict({
        'language': 'python',
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': '''
                def hello_world():
                    print("Hello World!")
  
              ''',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'object': 'block',
      'type': 'code',
    }),
    dict({
      'divider': dict({
      }),
      'object': 'block',
      'type': 'divider',
    }),
    dict({
      'object': 'block',
      'paragraph': dict({
        'rich_text': list([
          dict({
            'annotations': dict({
              'bold': False,
              'code': False,
              'color': 'default',
              'italic': False,
              'strikethrough': False,
              'underline': False,
            }),
            'text': dict({
              'content': '',
            }),
            'type': 'text',
          }),
        ]),
      }),
      'type': 'paragraph',
    }),
    dict({
      'object': 'block',
      'paragraph': dict({
        'rich_text': list([
          dict({
            'external': dict({
              'url': 'https://example.com/image.jpg',
            }),
            'type': 'image',
          }),
        ]),
      }),
      'type': 'paragraph',
    }),
    dict({
      'object': 'block',
      'table': dict({
        'children': list([
          dict({
            'table_row': dict({
              'cells': list([
                list([
                  dict({
                    'annotations': dict({
                      'bold': False,
                      'code': False,
                      'color': 'default',
                      'italic': False,
                      'strikethrough': False,
                      'underline': False,
                    }),
                    'text': dict({
                      'content': 'Header 1',
                    }),
                    'type': 'text',
                  }),
                ]),
                list([
                  dict({
                    'annotations': dict({
                      'bold': False,
                      'code': False,
                      'color': 'default',
                      'italic': False,
                      'strikethrough': False,
                      'underline': False,
                    }),
                    'text': dict({
                      'content': 'Header 2',
                    }),
                    'type': 'text',
                  }),
                ]),
              ]),
            }),
            'type': 'table_row',
          }),
          dict({
            'table_row': dict({
              'cells': list([
                list([
                  dict({
                    'annotations': dict({
                      'bold': False,
                      'code': False,
                      'color': 'default',
                      'italic': False,
                      'strikethrough': False,
                      'underline': False,
                    }),
                    'text': dict({
                      'content': 'Cell 1',
                    }),
                    'type': 'text',
                  }),
                ]),
                list([
                  dict({
                    'annotations': dict({
                      'bold': False,
                      'code': False,
                      'color': 'default',
                      'italic': False,
                      'strikethrough': False,
                      'underline': False,
                    }),
                    'text': dict({
                      'content': 'Cell 2',
                    }),
                    'type': 'text',
                  }),
                ]),
              ]),
            }),
            'type': 'table_row',
          }),
          dict({
            'table_row': dict({
              'cells': list([
                list([
                  dict({
                    'annotations': dict({
                      'bold': False,
                      'code': False,
                      'color': 'default',
                      'italic': False,
                      'strikethrough': False,
                      'underline': False,
                    }),
                    'text': dict({
                      'content': 'Cell 3',
                    }),
                    'type': 'text',
                  }),
                ]),
                list([
                  dict({
                    'annotations': dict({
                      'bold': False,
                      'code': False,
                      'color': 'default',
                      'italic': False,
                      'strikethrough': False,
                      'underline': False,
                    }),
                    'text': dict({
                      'content': 'Cell 4',
                    }),
                    'type': 'text',
                  }),
                ]),
              ]),
            }),
            'type': 'table_row',
          }),
        ]),
        'has_column_header': True,
        'has_row_header': False,
        'table_width': 2,
      }),
      'type': 'table',
    }),
  ])
# ---
